libs/ledgerjs/packages/react-native-hw-transport-ble/src/BleTransport.ts

```diff
@@ -685,19 +829,36 @@ export default class BleTransport extends Transport {
   /**
    * Do not call this directly unless you know what you're doing. Communication
    * with a Ledger device should be through the {@link exchange} method.
-   * @param buffer
-   * @param txid
+   *
+   * For each call a transaction id is added to the current stack of transaction ids.
+   * With this transaction id, a pending BLE communication operations can be cancelled.
+   * Note: each frame/packet of a longer BLE-encoded message to be sent should have their unique transaction id.
+   *
+   * @param buffer BLE-encoded packet to send to the device
+   * @param frameId Frame id to make `write` aware of a bigger message that this frame/packet is part of.
+   *  Helps creating related a collection of transaction ids
    */
-  write = async (buffer: Buffer, txid?: string | undefined): Promise<void> => {
+  write = async (buffer: Buffer): Promise<void> => {
+    const transactionId = uuid();
+    this.currentTransactionIds.push(transactionId);
+
+    const tracer = this.tracer.withUpdatedContext({ transactionId });
+    tracer.trace("Writing to device", { willMessageBeAcked: !this.writeCmdCharacteristic });
+
     try {
       if (!this.writeCmdCharacteristic) {
-        await this.writeCharacteristic.writeWithResponse(buffer.toString("base64"), txid);
+        // The message will be acked by the device
```

Comment:
is this availability of writing with response depending on the device ?
btw I'm wondering if we shouldn't use a wording closer to the react-native-ble-plx API, also not starting with a verb which to me means it's a function/method.
IMO it would improve greatly the readability for new eyes.
And those properties are only accessed in this file IMO so no risk of impacting other stuff.
Something like `characteristicWriteableWithResponse`/`characteristicWriteableWithoutResponse`.

Reply:
I would say that it's decided during the "hand-shake" when the BLE communication is started with the device. We should search and add it to the documentation we wanted to do.
But my understanding is that we know the UUID of a "write with response/ack" characteristic and the UUID of a "write without response" characteristic for each device (defined in `libs/ledgerjs/packages/devices/src/index.ts`, there is the "notify characteristic" too), and when the device gives us the service and the characteristics we can use to communicate with it, we check which one we have access to.
And we always check that the "write with response" characteristic (`writeC`) exists (also the "notify characteristic") in the `open` function.

Reply:
and for the name i agree with you

Reply:
Just maybe:
- `writeableWithResponseCharacteristic`
- `writeableWithoutResponseCharacteristic`

Reply:
perfect 👍